Removes acl files from list of files to match when mapping url to file

megoth opened the pull request:
Should fix https://github.com/solid/node-solid-server/issues/1104

megoth commented:
I'll push an update soon that doesn't hardcode it ^_^

megoth commented:
@rubensworks @kjetilk I've pushed an update, hope it looks better ^_^ (Was uncertain how to include the knowledge of suffixes into resourceMapper, so hope the way I did it looks ok)

rubensworks commented on lib/resource-mapper.js:
```diff
@@ -90,6 +90,14 @@ class ResourceMapper {
+
+    // Find a file with the same name (minus the dollar extension)
+    return (files.find(f => this._removeDollarExtension(f) === filename ||
+      (isIndex && f.startsWith(this._indexFilename + '.') && !f.endsWith(this._suffixes.acl))))
```
I would go one step further, and loop over all `this._suffixes` to exclude _all_ file names that end with any of the defined suffixes. Right now, only `index.html.acl` will be blacklisted, but `index.html.meta` will still cause the same problem. A loop over `this._suffixes` would solve this.

For this purpose, it may be easier to refactor `this._suffixes` to be an array instead of a hash, as this is (AFAICS) only being used here.

megoth commented:
@rubensworks then I've implemented your suggested changes (changing fileSuffixes from an object to an array). Could you have a look at the latest change, and see if you think it's good?

RubenVerborgh commented on lib/resource-mapper.js:
```diff
@@ -15,7 +15,8 @@ class ResourceMapper {
     indexFilename = 'index',
-    overrideTypes = { acl: 'text/turtle', meta: 'text/turtle' }
+    overrideTypes = { acl: 'text/turtle', meta: 'text/turtle' },
+    fileSuffixes = ['.acl', '.meta']
```
Unclear variable name. I would name it something like `controlExtensions`.

RubenVerborgh commented on lib/resource-mapper.js:
```diff
@@ -24,6 +25,7 @@ class ResourceMapper {
     this._defaultContentType = defaultContentType
     this._indexFilename = indexFilename
     this._types = { ...types, ...overrideTypes }
+    this._suffixes = fileSuffixes
```
Idem. They're not just suffixes, they are suffixes for a specific kinds of files that needs special behavior.

RubenVerborgh commented on lib/resource-mapper.js:
```diff
@@ -68,11 +70,9 @@ class ResourceMapper {
       // Read all files in the corresponding folder
```
Outdated comment

RubenVerborgh commented on lib/resource-mapper.js:
```diff
@@ -68,11 +70,9 @@ class ResourceMapper {
       // Find a file with the same name (minus the dollar extension)
-      let match = !searchIndex ? '' : (files.find(f => this._removeDollarExtension(f) === filename ||
-        (isIndex && f.startsWith(this._indexFilename + '.'))))
+      let match = searchIndex ? await this._getMatchingFile(folder, filename, isIndex) : ''
```
I'm not sure if I would move this specific bit to a new function. Is there potential for reuse?

megoth replied:
My thinking was to make it more readable by giving it a name. Don't know of any potential reuse, no.

RubenVerborgh commented on lib/resource-mapper.js:
```diff
@@ -90,6 +90,14 @@ class ResourceMapper {
+    const files = await this._readdir(folder)
+
+    // Find a file with the same name (minus the dollar extension)
+    return (files.find(f => this._removeDollarExtension(f) === filename ||
```
We should probably expand this condition across multiple lines now, given the increased complexity.

RubenVerborgh commented on lib/resource-mapper.js:
```diff
@@ -90,6 +90,14 @@ class ResourceMapper {
+
+    // Find a file with the same name (minus the dollar extension)
+    return (files.find(f => this._removeDollarExtension(f) === filename ||
+      (isIndex && f.startsWith(this._indexFilename + '.') && !this._suffixes.some(suffix => f.endsWith(suffix)))))
```
Seems wasteful to always iterate over the extensions.

How about, in the constructor, we instead use a regex to determine whether something is a control file? Then the default value is `/.(?:acl|meta)$/`, and this routine just does `this._controlFile.test(f)`.

megoth replied:
That sounds reasonable - I'll write tests that make sure stuff works, and then try to optimize if per your suggestions 😸

megoth commented:
Ah, yes, I've been lazy on tests here. I'll write them when I have time, hopefully during the next couple of days.

megoth commented:
@RubenVerborgh think I have addressed your concerns; I'm uncertain of how to run coverage reports on this actually, could you check? (And perhaps drop a line on how you do it here? Could read up more on it, but am in bit of a hurry right now =P )

RubenVerborgh commented on lib/resource-mapper.js:
```diff
@@ -24,6 +25,7 @@ class ResourceMapper {
     this._defaultContentType = defaultContentType
     this._indexFilename = indexFilename
     this._types = { ...types, ...overrideTypes }
+    this._isControlFile = new RegExp(`.(?:${fileSuffixes.join('|')})$`)
```
Not the right regex: a) fileSuffixes already has a leading dot, so there are two dots now b) dots need to be escaped in regexes (but, since you are creating the regex from a string, actually needs double backslash '`\\.`).

RubenVerborgh replied:
Alternatively, you can just pass the regex itself instead of passing fileSuffixes.

megoth replied:
That makes it difficult to configure the suffixes though =/

megoth replied:
Hope the latest commit looks good.

Ryuno-Ki commented:
If you're concerned about coverage... why not use a tool like [CodeCov](https://codecov.io/)?
You could make it even part of the conditions to flag a PR as mergeable.

RubenVerborgh commented on lib/resource-mapper.js:
```diff
@@ -24,6 +25,7 @@ class ResourceMapper {
     this._defaultContentType = defaultContentType
     this._indexFilename = indexFilename
     this._types = { ...types, ...overrideTypes }
+    this._isControlFile = new RegExp(`(?:${fileSuffixes.map(fs => fs.replace('.', '\\.')).join('|')})$`)
```
`fs.replace(/\./g, '\\.')` strictly, but good enough 😉

megoth replied:
Regex is so difficult =P

RubenVerborgh commented:
> If you're concerned about coverage... why not use a tool like [CodeCov](https://codecov.io/)?

Coverage in general is not great yet; but it was for this particular class 😎
